perf: optimize srcset functions

baldurh opened the pull request:
## Description

Optimizes the srcset generating functions by eliminating template strings since they are transpiled to less performing `String.prototype.concat()` and by eliminating dpr quality constant values.


Benchmarks in ms. 100,000 iterations

|                                       | before   | after    | change  |
|---------------------------------------|----------|----------|---------|
| responsive                            | 1670.436 | 1575.341 | -5.69%  |
| responsive with fixed height          | 1345.359 | 1263.620 | -6.08%  |
| responsive with aspect ratio          | 1762.956 | 1667.487 | -5.42%  |
| fixed width                           | 1366.371 | 1221.589 | -10.60% |
| fixed width with quality              | 1363.528 | 1221.570 | -10.41% |
| fixed width with dpr quality disabled | 1310.299 | 1209.029 | -7.73%  |
| fixed width with aspect ratio         | 1336.828 | 1205.730 | -9.81%  |


With latest changes

|                                       | before   | after    | change  |
|---------------------------------------|----------|----------|---------|
| responsive                            | 1766.633 | 1551.187 | -12.20% |
| responsive with fixed height          | 1661.339 | 1339.880 | -19.35% |
| responsive with aspect ratio          | 1999.471 | 1719.327 | -14.01% |
| fixed width                           | 1582.983 | 1297.991 | -18.00% |
| fixed width with quality              | 1588.295 | 1376.398 | -13.34% |
| fixed width with dpr quality disabled | 1394.533 | 1246.995 | -10.58% |
| fixed width with aspect ratio         | 1457.520 | 1268.894 | -12.94% |

[Flamegraph master](https://upload.clinicjs.org/public/b03b7077def184b36074ab9ef9bd79fd8c807c5baedb6725f74c3ec5f759f387/28561.clinic-flame.html#selectedNode=241&zoomedNode=&exclude=83c0&merged=true)
[Flamegraph this branch](https://upload.clinicjs.org/public/6f89e04f36a87f97b617d5b88f8aea0a59fd2e36181b055519630f7e5fad1aa4/28891.clinic-flame.html#selectedNode=188&zoomedNode=&exclude=83c0&merged=true)

frederickfogerty commented on src/react-imgix.jsx:
```diff
@@ -91,14 +99,19 @@ function buildSrcSetPairFunction(url, aspectRatioDecimal = 0, fixedHeight) {
-  if (disableQualityByDPR && quality) {
-    return dpr => `${url}&q=${quality}&dpr=${dpr} ${dpr}x`;
+  if (quality) {
+    const buildDprSrcWithQuality = (_, dpr) =>
```
What I've gleaned from modern Node perf blog posts is that this could potentially perform even better if this is defined statically outside the `buildDprSrcFunction` function, because Node will automatically memoize/optimize the function.

sherwinski replied:
@frederickfogerty Mind sharing some of those blog posts? 😅

frederickfogerty commented on src/react-imgix.jsx:
```diff
@@ -143,7 +156,9 @@ function buildSrc({
         q
       );
-      srcSet = [1, 2, 3, 4, 5].map(buildDprSrc).join(", ");
+      srcSet = Object.values(CONSTANTS)
```
How does the performance change if this is replaced with a for loop?

baldurh commented:
@frederickfogerty thanks for your suggestions! I tried them out and saw some overall performance improvements so I added them here. The benchmarks seem to suggest some of the tests actually perform worse but the average got better. I’ve updated the description with the latest numbers.

frederickfogerty commented:
> I tried them out and saw some overall performance improvements so I added them here.

That's great to hear. Glad I could help out 👍

baldurh commented:
> This LGTM as well 👍
> I do want to note that with imgix's newly-added `ar` parameter, I think a lot of the aspect ratio work in this project can be refactored out. But that's a discussion for a different time.

@sherwinski That’s great since the function calculating the height using aspect ratio is the heaviest one of the srcSet functions 🥳
